src/testing/slim_check/testable.lean

```diff
@@ -0,0 +1,490 @@
+data to test the proposition `p`.  For instance, `testable.check (∀ x
+: ℕ, 2 ∣ x → x < 100)` builds the `testable` instance step by step
+with:
+
+```
+testable (∀ x : ℕ, 2 ∣ x → x < 100) -: sampleable ℕ, decidable (λ x, 2 ∣ x), testable (λ x, x < 100)
+testable (λ x, x < 100)              -: decidable (λ x, x < 100)
+```
+
+`sampleable ℕ` lets us create random data of type `ℕ` in a way that
+helps find small counter-examples.  Next, the test of the proposition
+hinges on `2 ∣ 100` and `x < 100` to both be decidable. The
+implication between the two could be tested as a whole but it would be
+less informative. Indeed, we could generate a hundred odd numbers and
+the property would be shown to hold for each but the right conclusion
+is that we haven't found meaningful examples. Instead, when `2 ∣ x`
+does not hold, we reject the example (i.e.  we do not count it toward
+the 100 required positive examples) and we start over. Therefore, when
+`testable.check` prints `Success`, it means that a hundred even
+numbers were found to be less than 100.
+
+### Polymorphism
+
+The property `testable.check (∀ (α : Type) (xs ys : list α), xs ++ ys
+= ys ++ xs)` shows us that type-polymorphic properties can be
+tested. `α` is instantiated with `ℤ` first and then tested as normal
+monomorphic properties.
+
+### What do I do if I'm testing a property about my newly defined type?
+
+Let us consider a type made for a new formalization:
+
+```lean
+structure my_type :=
+(x y : ℕ)
+(h : x ≤ y)
+```
+
+How do we test a property about `my_type`? For instance, let us consider
+`testable.check $ ∀ a b : my_type, a.y ≤ b.x → a.x ≤ b.y`. Writing this
+property as is will give us an error because we do not have an instance
+of `sampleable my_type`. We can define one as follows:
+
+```lean
+instance : sampleable my_type :=
+{ sample := do
+  x ← sample ℕ,
+  xy_diff ← sample ℕ,
+  return { x := x, y := x + xy_diff, h := /- some proof -/ } }
+```
+
+We can see that the instance is very simple because our type is built
+up from other type that have `sampleable` instances. `sampleable` also
+has a `shrink` method but it is optional. We may want to implement one
+for ease of testing as:
+
+```lean
+/- ... -/
+  shrink := λ ⟨x,y,h⟩, (λ ⟨x,y⟩, { x := x, y := x + y, h := /- proof -/}) <$> shrink (x, y - x)
+}
+```
+
+Again, we take advantage of the fact that other types have useful
+`shrink` implementations, in this case `prod`.
+
+## Main definitions
+  * `testable` class
+  * `testable.check` a way to test a proposition using random examples
+
+## Tags
+
+random testing
+
+## References
+
+  * https://hackage.haskell.org/package/QuickCheck
+
+-/
+
+universes u v
+
+variable α : Type u
+variable β : α → Prop
+variable f : Type → Prop
+
+namespace slim_check
+
+/-- Result of trying to disprove `p` -/
+@[derive inhabited]
+inductive test_result (p : Prop)
+  /- succeed when we find another example satisfying `p` -/
+| success : (psum unit p) → test_result
```

Comment:
Please explain what the parameters to these constructors represent.

Reply:
Thanks. Could you move the explanations to the doc string so they appear in the docs?